### PR TITLE
enable the usage of Netty OpenSSL provider implementation and PKCS12 key stores

### PR DESCRIPTION
**Pull request by mscheibler**
In order to enable the usage of additional TLS cyphers which are not supported by the JDK SSL implementation it would be nice to have the option to use the Netty OpenSSL provider implementation.

Also the use of PKCS12 key stores is now supported.

The defaults are set to match the previous behavior.

### CODE REVIEW

**Comment by mscheibler** [comment]
Hi @andsel 
If you could spare a few minutes it would be good to know whether my change to ISslContextCreator gets your approval. If not, then i would have to think of another approach.
Thanks a lot.

**Comment by andsel** [comment]
Hi Mike, I've just to find the time to review it.

File: broker/src/main/java/io/moquette/server/DefaultMoquetteSslContextCreator.java [hunk 1]
```diff
@@ -38,77 +53,120 @@
...
         if (jksPath == null || jksPath.isEmpty()) {
-            // key_store_password or key_manager_password are empty
             LOG.warn("The keystore path is null or empty. The SSL context won't be initialized.");
             return null;
         }
-
-        // if we have the port also the jks then keyStorePassword and keyManagerPassword
-        // has to be defined
         final String keyStorePassword = props.getProperty(BrokerConstants.KEY_STORE_PASSWORD_PROPERTY_NAME);
-        final String keyManagerPassword = props.getProperty(BrokerConstants.KEY_MANAGER_PASSWORD_PROPERTY_NAME);
```
Comment (andsel) [inline broker/src/main/java/io/moquette/server/DefaultMoquetteSslContextCreator.java:61]: Why the key manager password has been removed,it's not anymore necessary?

File: broker/src/main/java/io/moquette/server/netty/NettyAcceptor.java [hunk 2]
```diff
@@ -86,9 +101,9 @@ protected void encode(ChannelHandlerContext chc, ByteBuf bb, List<Object> out) t
         }
     }
 
-    abstract static class PipelineInitializer {
+    private abstract static class PipelineInitializer {
 
-        abstract void init(ChannelPipeline pipeline) throws Exception;
```
Comment (andsel) [inline broker/src/main/java/io/moquette/server/netty/NettyAcceptor.java:91]: Why this change,  PipelineInitializer speacks about a ChannelPipeline, why moving to SocketChannel?

**Review by andsel — COMMENTED** [review]
I think that's almost ok. It's not clear to me why in PipileneInitializer you switched the argument of init from ChannelPipeline to SocketChannel, considering that you then retrieve the ChannelPipeline from SocketChannel itself
Then please add a section in the distribution/src/main/resources/moquette.conf with the various option of configuration for the SSL handler, with line commented.

File: broker/src/main/java/io/moquette/server/DefaultMoquetteSslContextCreator.java (hunk 1 above)
Comment (mscheibler) [reply broker/src/main/java/io/moquette/server/DefaultMoquetteSslContextCreator.java:61]: The `KEY_MANAGER_PASSWORD_PROPERTY_NAME` is still used to initialize the KeyManagerFactory (when the JDK provider is used) and also to access the private key (for handover to the OpenSSL provider). However i did rename the variable to `keyPassword` as i think this describes the purpose of this password better. According to the Javadoc it is *the password for recovering keys in the KeyStore*
Just the `loadKeyStore()` method does not need it.

File: broker/src/main/java/io/moquette/server/netty/NettyAcceptor.java (hunk 2 above)
Comment (mscheibler) [reply broker/src/main/java/io/moquette/server/netty/NettyAcceptor.java:91]: Because i handover the `SocketChannel` to the `createSslHandler` method.
It would also be possible to retrieve the channel from the pipeline, but then i would need to add a cast, as the pipeline just returns a `Channel` not a `SocketChannel` and i wanted to avoid that.
The channel information about the remote peer is then used to initialize the SSL engine.

**Comment by andsel** [comment]
Thanks @mscheibler
